wcfsetup/install/files/lib/acp/form/LanguageAddForm.class.php

```diff
@@ -103,149 +75,123 @@ public function readParameters()
+                        ->description('wcf.acp.language.name.description')
+                        ->maximumLength(255)
+                        ->required(),
+                    TextFormField::create('languageCode')
+                        ->label('wcf.acp.language.code')
+                        ->description('wcf.acp.language.code.description')
+                        ->maximumLength(20)
+                        ->required()
+                        ->addValidator(new FormFieldValidator('unique', function (TextFormField $formField) {
+                            if ($formField->getValidationErrors() !== []) {
+                                return;
+                            }
+
+                            if (
+                                $this->formObject instanceof Language
+                                && \mb_strtolower($this->formObject->languageCode) === $formField->getValue()
+                            ) {
+                                return;
+                            }
+
+                            if (LanguageFactory::getInstance()->getLanguageByCode($formField->getValue())) {
+                                $formField->addValidationError(new FormFieldValidationError(
+                                    'notUnique',
+                                    'wcf.acp.language.add.languageCode.error.notUnique'
+                                ));
+                            }
+                        }))
+                        ->addValidator(new FormFieldValidator(
+                            'locale',
+                            function (SingleSelectionFormField $formField) {
+                                if ($formField->getValidationErrors() !== []) {
+                                    return;
+                                }
+
+                                $localeField = $formField->getDocument()->getNodeById('locale');
+                                \assert($localeField instanceof SingleSelectionFormField);
+                                if ($localeField->getValue() === '') {
+                                    return;
+                                }
+
+                                $test = \explode('_', $localeField->getValue());
+                                if (\count($test) === 1 || $formField->getValue() !== $test[0]) {
+                                    $formField->addValidationError(
+                                        new FormFieldValidationError(
+                                            'localeMismatch',
+                                            'wcf.acp.language.add.languageCode.error.localeMismatch',
+                                            [
+                                                'locale' => $localeField->getValue(),
+                                            ]
+                                        )
+                                    );
+                                }
+                            }
+                        )),
```

Comment:
I'm afraid I'm getting an exception here whenever I submit the form.

```
wcf\acp\form\LanguageAddForm::wcf\acp\form\{closure}(): Argument #1 ($formField) must be of type wcf\system\form\builder\field\SingleSelectionFormField, wcf\system\form\builder\field\TextFormField given, called in /var/www/html/lib/system/form/builder/field/validation/FormFieldValidator.class.php on line 71
```

The validator needs to be attached to the `locale` field, not the `languageCode` field.

Reply:
Ah sorry, I moved the validator and didn't fix that up.

Reply:
Since the user generated content is the language code, the language code should be shown as faulty, not the system given locale the user wouldn't be able to change.

Reply:
It's a mismatch between both fields, so either would be an acceptable error target. However the select field comes later, so the issue is that the selection does not match the text input and the error should be attached to the select. Especially since the locale (region) selection is optional.

Please have a look at the proposed validator, I've tested it and it appears to match the desired behavior.

Reply:
I can move the Validator in a follow-up commit to save you some work, as the current one is at least functional, but the Exception when adding a language is a blocker (see the other comment). We want to avoid broken commits as much as possible.